src/memray/_memray/snapshot.cpp

```diff
@@ -407,6 +420,163 @@ HighWaterMarkAggregator::visitAllocations(const allocation_callback_t& callback)
+            }
+            break;
+        }
+        case hooks::AllocatorKind::RANGED_ALLOCATOR: {
+            const Allocation& allocation = allocation_or_deallocation;
+            size_t generation = d_num_snapshots;
+            d_mmap_intervals.addInterval(
+                    allocation.address,
+                    allocation.size,
+                    {std::make_shared<Allocation>(allocation), generation});
+            break;
+        }
+        case hooks::AllocatorKind::RANGED_DEALLOCATOR: {
+            const Allocation& deallocation = allocation_or_deallocation;
+            auto removal_stats =
+                    d_mmap_intervals.removeInterval(deallocation.address, deallocation.size);
+            for (const auto& [interval, pair] : removal_stats.freed_allocations) {
+                recordRangedDeallocation(pair.first, interval.size(), pair.second);
+            }
+            for (const auto& [interval, pair] : removal_stats.shrunk_allocations) {
+                recordRangedDeallocation(pair.first, interval.size(), pair.second);
+            }
+            for (const auto& [interval, pair] : removal_stats.split_allocations) {
+                recordRangedDeallocation(pair.first, interval.size(), pair.second);
+            }
+            break;
+        }
+    }
+}
+
+HighWaterMarkLocationKey
+AllocationLifetimeAggregator::extractKey(const Allocation& allocation)
+{
+    return {allocation.tid,
+            allocation.frame_index,
+            allocation.native_frame_id,
+            allocation.native_segment_generation,
+            allocation.allocator};
+}
+
+void
+AllocationLifetimeAggregator::recordRangedDeallocation(
+        const std::shared_ptr<Allocation>& allocation_ptr,
+        size_t bytes_deallocated,
+        size_t generation_allocated)
+{
+    // We hold one reference, and the IntervalTree may or may not hold others.
+    // We use a count of 0 for all but the last deallocation of a range so that
+    // partial deallocations won't affect the count of allocations by location.
+    bool fully_deallocated = allocation_ptr.use_count() == 1;
+    recordDeallocation(
+            extractKey(*allocation_ptr),
+            (fully_deallocated ? 1 : 0),
+            bytes_deallocated,
+            generation_allocated);
+}
+
+void
+AllocationLifetimeAggregator::recordDeallocation(
+        const HighWaterMarkLocationKey& key,
+        size_t count_delta,
+        size_t bytes_delta,
+        size_t generation)
+{
+    if (d_num_snapshots == generation) {
+        // Allocated and deallocated within the same snapshot. We can ignore this.
+        return;
+    }
+
+    auto& counts = d_allocation_history[std::make_tuple(generation, d_num_snapshots, key)];
+    counts.first += count_delta;
+    counts.second += bytes_delta;
+}
+
+void
+AllocationLifetimeAggregator::captureSnapshot()
+{
+    ++d_num_snapshots;
+}
+
+std::vector<AllocationLifetime>
+AllocationLifetimeAggregator::generateIndex()
```

Comment:
Can we add a small comment with the general recipe for this? I think is clear enough without comments here but the fact that we have some like

```
// Leaked simple allocations
```

makes it seems the reader should now already what the process is going to be and is a bit confusing the first time you go through this. WDYT?

Reply:
I tried to improve the comments here. Take a look and see if you think there's anything that still needs to be explained better.